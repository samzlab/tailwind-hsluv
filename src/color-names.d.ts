import { hexColor } from "./utils";

import { hexColor } from './utils.d.ts';

export type colorPair = [string, string];
export type colorNames = Array<colorPair>;

export function hexFromColorName(name: string): hexColor;

export default colorNames;