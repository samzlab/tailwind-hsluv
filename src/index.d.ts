type hexColor = string;
type namedColor = string;
// enum namedColor {
//     MacaroniAndCheese = "Macaroni and Cheese"
// };

type colorTuple = [number, number, number];

interface generateOptions {
    step?: number = 100,
    hpluv?: boolean
};

type inputColor = hexColor | namedColor;

interface inputColors {
    [key: string]: inputColor
};

type colorVariations = {
    [lightness: string]: hexColor
}

type outputColors = {
    [key: string]: colorVariations
}

type generatorFunction = (lightness: number) => hexColor;

export function fromRGB(r: number, g: number, b: number, hpluv?: boolean = false): generatorFunction;

export function colorFromName(name: inputColor): colorTuple;

export function generateSteps(generator: generatorFunction, step: number): colorVariations

export function generateColors(colors: inputColors, options?: generateOptions): outputColors