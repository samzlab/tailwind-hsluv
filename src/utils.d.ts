export type hslColor = [number, number, number];
export type rgbColor = [number, number, number];
export type hexColor = string;

export type generatorFunction = (lightness: number) => hexColor;

export function rgbToHsl(rgb: rgbColor): hslColor;

export function createGenerator(rgb: rgbColor, hpluv: boolean = false): generatorFunction;

export function hex2rgb(hex: hexColor): rgbColor;

export function rgb2hex(rgb: rgbColor): hexColor;