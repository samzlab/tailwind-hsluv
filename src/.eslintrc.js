module.exports = {
	parserOptions: {
		ecmaVersion: 2020, // 11
		sourceType: 'module'
	},
	env: {
		browser: true,
		es6: true
	},
	extends: [
		// add more generic rulesets here, such as:
		// 'eslint:recommended',
		'@koffeine',
		'plugin:vue/vue3-recommended'
	],
	rules: {
		// override/add rules settings here, such as:
		// 'vue/no-unused-vars': 'error'
		'vue/no-multiple-template-root': 'off',
		'vue/valid-template-root': 'off',
		'vue/script-indent': [ 'error', 4, { baseIndent: 1 } ],
		'vue/html-indent': [ 'error', 4, { baseIndent: 1 } ],
		'vue/max-attributes-per-line': 'off',
		'no-bitwise': 'off',
		'func-style': 'off'
	},
	overrides: [
		{
			files: [ '*.vue' ],
			rules: {
				indent: 'off'
			}
		}
	]
};