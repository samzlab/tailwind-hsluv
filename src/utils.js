import { hpluvToHex, hsluvToHex, rgbToHsluv } from 'hsluv';
import assert from 'assert';

/**
 *
 * @param {import('../types/utils').rgbColor} rgb
 * @param {boolean} hpluv
 *
 * @returns { import('../types/utils').generatorFunction }
 */
export function createGenerator(rgb, hpluv = false) {
	const [ hue, saturation ] = rgbToHsluv(rgb);
	const method = hpluv ? hpluvToHex : hsluvToHex;

	return (lightness) => {
		assert(typeof lightness === 'number', 'Lightness value must be a number');
		assert(lightness >= 0 && lightness <= 100, 'Lightness value must be between 0-100)');

		return method([ hue, 100 - saturation, lightness ]);
	};
}

/**
 * @param {string} hex
 */
export function normalizeHex(hex) {
	/** @type {string[]} */
	let parts = [];
	let input = hex;

	hex = hex.toUpperCase();

	if (hex.charAt(0) === '#') {
		hex = hex.slice(1);
	}

	assert(hex.match(/^[0-9A-F]{3,6}$/ui), `Invalid hex color value: ${input}`);

	if (hex.length === 3) {
		parts = hex.match(/[0-9A-F]/gu).map((part) => `${part}${part}`);
	} else if (hex.length === 6) {
		parts = hex.match(/[0-9A-F]{2}/gu);
	}

	assert(parts.length === 3, `Invalid hex color length: ${input}`);

	return parts.join('');
}

/**
 * @param {import('../types/utils').hexColor} hex
 * @returns {import('../types/utils').rgbColor}
 */
export function hex2rgb(hex) {
	const parts = normalizeHex(hex).match(/[0-9A-F]{2}/gu);

	assert(parts.length === 3, `Invalid hex color: ${hex}`);

	const rgb = parts.map((hexValue) => parseInt(hexValue, 16));

	assert(!rgb.find((part) => part < 0 || part > 255), `Hex color (${hex}) resolved to invalid rgb(${rgb.join(',')})`);

	// @ts-ignore
	return rgb;
}

/**
 * @param {import('../types/utils').rgbColor} rgb
 * @returns {import('../types/utils').hexColor}
 */
export function rgb2hex(rgb) {
	return rgb.map((int) => int.toString(16).padStart(2, '0')).join('').toUpperCase();
}