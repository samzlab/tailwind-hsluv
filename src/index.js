import { rgb2hex, hex2rgb, createGenerator, normalizeHex } from './utils.js';
import { hexFromColorName } from './color-names';

/**
 *
 * Generate color variants by hsluv color generator function for each step evenly
 *
 * @param { import('../types/utils').generatorFunction } generator
 * @param { number } step
 *
 */
export function generateSteps(generator, step = 100) {
	const result = {};

	for (let i = step; i < 1000; i += step) {
		result[`${i}`] = generator(100 - (i / 10));
	}

	return result;
}

/**
 * Resolve a inputColor string or array to rgb/hex values
 *
 * @param {import('../types/index').inputColor} color
 */
export function resolveColor(color) {
	/** @type {import('../types/utils').rgbColor} */
	let rgb;

	/** @type {import('../types/utils').hexColor} */
	let hex;

	if (Array.isArray(color)) {

		rgb = color;
		hex = rgb2hex(rgb);

	} else if (typeof color === 'string') {

		if (color.charAt(0) === '#') {
			hex = normalizeHex(color);
		} else {
			hex = hexFromColorName(color);
		}

		rgb = hex2rgb(hex);

	} else {

		throw Error(`Invalid color value: ${color} (expected String or Array instead of ${typeof color})`);

	}

	return { rgb, hex };
}

/**
 *
 * Generate color variants for TailwindCSS config by color map
 *
 * @param { import('../types/index').inputColorsMap } colors
 * @param { import('../types/index').generateOptions } options
 *
 */
export function generateColors(colors, { step = 100, hpluv = false } = {}) {
	return Object.entries(colors).reduce((result, [ key, /** @type { import('../types/index').inputColor } */ inputColor ]) => {
		const { rgb, hex } = resolveColor(inputColor);

		result[key] = generateSteps(createGenerator(rgb, hpluv), step);
		result[key].default = `#${hex}`;

		return result;
	}, {});
}