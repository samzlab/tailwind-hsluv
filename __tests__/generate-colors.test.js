import { generateColors, resolveColor } from '../src';
import { AssertionError } from 'assert';

describe('Generate colors', () => {

	test('"red" should resolve to "#ff0000"', () => {
		expect(resolveColor('red')).toEqual({ hex: 'FF0000', rgb: [ 255, 0, 0 ] });
	});

	test('color variations has default key', () => {
		expect(generateColors({
			red: 'red'
		})).toHaveProperty([ 'red', 'default' ]);
	});

	test('variations (step=50) contains 50...950', () => {
		const colors = generateColors({ red: 'red' }, { step: 50 });

		expect(colors).toHaveProperty([ 'red', '50' ]);
		expect(colors).toHaveProperty([ 'red', '950' ]);
	});

	test('invalid step values (510, 45) should throw Error', () => {
		const colorMap = { red: 'red' };

		expect(() => generateColors(colorMap, { step: 510 })).toThrow(AssertionError);
		expect(() => generateColors(colorMap, { step: 45 })).toThrow(AssertionError);
	});

	test('should generate "red" colors', () => {
		expect(generateColors({ red: 'red' })).toMatchSnapshot();
	});

	test('should generate HPLuv colors', () => {
		expect(generateColors({ red: 'red' }, { hpluv: true })).toMatchSnapshot();
	});


});