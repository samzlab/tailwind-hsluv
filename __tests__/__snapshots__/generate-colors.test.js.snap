// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate colors should generate "red" colors 1`] = `
Object {
  "red": Object {
    "100": "#FFDADA",
    "200": "#FFB3B3",
    "300": "#FF8888",
    "400": "#FF5353",
    "500": "#EF0000",
    "600": "#C00000",
    "700": "#930000",
    "800": "#690000",
    "900": "#410000",
    "default": "#FF0000",
  },
}
`;

exports[`Generate colors should generate HPLuv colors 1`] = `
Object {
  "red": Object {
    "100": "#FEDADA",
    "200": "#FDB4B4",
    "300": "#E69696",
    "400": "#C37E7E",
    "500": "#A16868",
    "600": "#815252",
    "700": "#623D3D",
    "800": "#442929",
    "900": "#291717",
    "default": "#FF0000",
  },
}
`;
